### DISCUSSION TITLE
How to deploy a NodeJS app in production mode?

### DISCUSSION THREAD

**Discussion by kamzata** [dokku/dokku#4934] [Q&A]
I successfully deployed my NodeJS app but it's still in developer mode. How can I set Dokku in order to get my app in production mode?

**Comment by Cellane** [comment]
Dokku has nothing to do with this. You need to set your app to production mode according to the standard practices of your platform, which in the case of Node.js might be [configuring the `NODE_ENV` environmental variable](https://nodejs.dev/learn/nodejs-the-difference-between-development-and-production). In the context of Dokku, that could be accomplished with `dokku config:set <app-name> NODE_ENV=production`.

**Reply by kamzata** [reply]
Should I set a buildpack on Dokku for my Nodejs app?

**Reply by Cellane** [reply]
I’m not sure what’s the usual recommendation, but I only set my own buildpack (to disable Dokku’s/Herokuish’s buildpack auto-detection) if the auto-detection fails.

**Reply by kamzata** [reply]
In my local environment I just run _npm run build_, then _serve -s build_ to run my app in production without issues.

Anyway, I already set the NODE_ENV variable to production inside my Docker containers, but my app is still running in developer mode.

Should I use [this approach](https://stackoverflow.com/a/58829484)?

After _git push dokku master_ is there something else which should I do?

**Reply by kamzata** [reply]
This is the _printenv_ output:

```
herokuishuser@3b9d4fec06bd:~$ printenv
DOKKU_PROXY_PORT=80
DOKKU_APP_TYPE=herokuish
HOSTNAME=3b9d4fec06bd
MEMORY_AVAILABLE=14336
CACHE_PATH=/cache
HEROKUISH_VERSION=0.5.31
DOKKU_APP_RESTORE=1
PWD=/app
PORT=5000
NODE_ENV=production
NODE_HOME=/app/.heroku/node
HOME=/app
TERM=xterm
USER=herokuishuser
SHLVL=1
CURL_CONNECT_TIMEOUT=90
CURL_TIMEOUT=600
SELF_EXECUTABLE=/usr/bin/herokuish
WEB_MEMORY=512
WEB_CONCURRENCY=28
DOKKU_PROXY_PORT_MAP=http:80:5000
PS1=${debian_chroot:+($debian_chroot)}\u@\h:\w\$
STACK=heroku-20
PATH=/app/.heroku/node/bin:/app/.heroku/yarn/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/app/bin:/app/node_modules/.bin
SELF=/exec
GIT_REV=623b6cb20fadf204b28c51ae7a2476007bf53d58
DYNO=web.1
DEBIAN_FRONTEND=noninteractive
OLDPWD=/
_=/usr/bin/printenv
```

This is my _package.json_ file:

```
...
    "scripts": {
        "start": "react-scripts start",
        "build": "react-scripts build",
        "test": "react-scripts test",
        "eject": "react-scripts eject"
    },
...
```

**Reply by kamzata** [reply]
I have been able to deploy the app in production mode by specifying this (Heroku) buildpack created for `create-react-app` :
```
dokku buildpacks set myapp https://github.com/mars/create-react-app-buildpack.git
```

and then add, in the app's root, the `Procfile` with this content:

```
web: bin/boot
```

However, even if it's set in production, the Javascript source code is not minified. Any thoughts?

**Reply by josegonzalez** [reply]
Sounds like your build command isn't minifying. What enables minification?

**Reply by kamzata** [reply]
> Sounds like your build command isn't minifying. What enables minification?

I run `react-scripts build` but, to be honest, I now noticed it doesn't seem minified even in the local build.
